### ISSUE TITLE
Find out how to connect to Bluetooth devices within Xamarin in order to assign label to BLE devices in proximity

### ISSUE THREAD

**Issue by Meisterlu** [amosproj/amos2022ss05-find-my-hearing-aid#7]
## User story
1. As a user
2. I want / need to track up to two BLE devices
3. So that I can detect the devices and can assign labels

## Acceptance criteria
- [x] * Detection of the devices
- [x] * Track location of two devices
- [x] * Assign labels to devices such as "hearing aid"
- [x] * research regarding signal strength

## Definition of done (DoD)
* TBD

**Comment by Chippelius** [comment]
### A summary of my findings 

It seems that, while it is possible to access the Bluetooth-functionality of both platforms in the respective platform-specific projects (FindMyBLEDevice.Android and FindMyBLEDevice.iOS), Xamarin does not provide a platform-independent Interface for use in the core project (FindMyBLEDevice). There are, however, independent libraries that add this independence. 
This means, we have two options: 
* Implement the Bluetooth-functionality we need for both platforms individually, with a common interface, and access this interface from the core project. (This should be what these libraries do.)
* Use a library that does this. 

The following is a list of such possible libraries, each accompanied by an article that uses it in an example. 
**_Note that even if a library in this list is eliminated from being potentially used by us, each article is still worth reading in order to gain a better understanding of how Bluetooth works._** 
* Library: https://github.com/aritchie/bluetoothle 
  * Article: https://servocode.com/blog/servocode.com/blog/bluetooth-le-for-xamarin.forms-how-to-control-things-from-the-level-of-mobile-app 
  * License: MIT 
  * Note: This library is currently moving to the Shiny Framework (next library below), support currently remains, but will probably end sometime in the future 
* Library: https://github.com/shinyorg/Shiny 
  * Article: https://blog.noser.com/bluetooth-low-energy-in-xamarin-app-ein-projektbericht/ 
  * License: MIT 
  * Note: I'm not 100% sure whether this library is still in beta or fully released. It does however also come with Location-services (GPS, Geofencing, & Activity Monitoring), so we could hit two birds with one stone. 
* Library: https://github.com/tbrushwyler/Xamarin.BluetoothLE 
  * Article: https://medium.com/enigmamx/how-to-integrate-bluetooth-le-in-xamarin-forms-2dcdf974703a 
  * License: MIT 
* Library: https://github.com/xabre/xamarin-bluetooth-le 
  * Article: https://www.jenx.si/2020/08/13/bluetooth-low-energy-uart-service-with-xamarin-forms/ 
  * License: Apache-2.0 

Concerning how to measure the signal strength: 
This is indicated via the Received Signal Strength Indicator (RSSI). The RSSI-value is a real number in decibel and is generally negative, meaning that the signal is weaker at the receiver than it was at the sender (0 would mean perfect transmission, positive would mean it is being amplified). 
The RSSI value can be measured not only for connected devices, but for every device that is visible to you: When scanning for devices in your vicinity, you send a broadcast-message out into the ether and every device that wants to be found sends a response message back to you. Through this response message, the RSSI value and therefore the signal strength to all of these visible devices is known. 
Note, however, that the RSSI value may be more or less useful or even be unavailable, depending on the device! (see: https://stackoverflow.com/questions/2149776/android-irc-office-hours-question-about-android-bluetooth-rssi) 
I haven't checked yet which of the above libraries allow reading the RSSI value, but I would be surprised if any of them don't. 

tl;dr: 
Xamarin does not have a platform-independent way of using bluetooth, so we can either write 2 different versions of platform-specific code, or use a library that adds a platform-independent interface. There are several options for such libraries. 
Signal strength can be measured via the RSSI-value.
